docs/new-resource.md

```diff
@@ -1,149 +1,215 @@
-for _, bucket := range response.Buckets {
-    b := *bucket
-    s.runner.Run(func() error {
-        return s.readBucket(b, results)
-    })
+type GithubRepository interface {
+	ListRepositories() ([]string, error)
+	ListTeams() ([]Team, error)
+	ListMembership() ([]string, error)
+	ListTeamMemberships() ([]string, error)
+	ListBranchProtection() ([]string, error)
 }
-if err := s.runner.Wait(); err != nil {
-    return nil, err
+
+type githubRepository struct {
+	client GithubGraphQLClient
+	ctx    context.Context
+	config githubConfig
+	cache  cache.Cache
 }
-```
 
-- an instance of `terraform.ResourceReader` that you can use to read resource using the supplier:
+func NewGithubRepository(config githubConfig, c cache.Cache) *githubRepository {
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: config.Token},
+	)
+	oauthClient := oauth2.NewClient(ctx, ts)
+
+	repo := &githubRepository{
+		client: githubv4.NewClient(oauthClient),
+		ctx:    context.Background(),
+		config: config,
+		cache:  c,
+	}
 
-```go
-s3Bucket, err := s.reader.ReadResource(aws.AwsS3BucketResourceType, name)
-if err != nil {
-    logrus.Warnf("Error reading bucket %s[%s]: %+v", name, aws.AwsS3BucketResourceType, err)
-    return err
+	return repo
 }
-appendValueIntoMap(results, aws.AwsS3BucketResourceType, s3Bucket)
 ```
 
-- an instance of the cloud provider SDK that you will use to retrieve resources list
+So as you can see this contains the logic to create the github client (it might be created outside the repository if it
+makes sense to share it between multiple repositories). It also get a cache so every request is cached.
+Driftctl sometimes needs to retrieve list of resources more than once, so we cache every request to avoid unnecessary call.
+
+### Enumerator
 
-### Deserializer
+This is used to build a resources list. Enumerators can be found in `pkg/remote/<providername>/<type>_enumerator.go`. It will call the cloud provider SDK to get the list of resources.
 
-The deserializer is used when reading resource from the terraform provider or from the state.
-The interface contains a `Deserialize(values []cty.Value) ([]resource.Resource, error)` method that you'll implement.
+Note that at this point resources should not be entirely fetched.
+Most of the resource returned by enumerator have empty attributes: they only represent type and terraform id.
 
-You should then deserialize the obtained cty values into your resource and return the list.
+**There are exception to this**:
+- Sometime, you will need some more information about resources to retrieve them using the provider they should be added to the resource attribute maps.
+- For some more complex cases, middleware needs more information that the id and type and in order to make classic run of driftctl coherent with a run with deep mode activated,
+these informations should be fetched manually by the enumerator using the remote sdk.
```

Comment:
`For complex cases (e.g. middlewares) where you would need driftctl to run as expected in deep and non-deep mode, you would need to enumerate resources as well as to fetch manually specific attributes, using the remote SDK, before adding them to the map of data.`